### PR TITLE
Merge some simple classes' type stubs back to source 

### PR DESCRIPTION
**Pull request by trim21**
This is a split of #2570, https://github.com/PyGithub/PyGithub/pull/2570 is too big, this make is easy to review.

### CODE REVIEW

File: github/HookDescription.py [hunk 1]
```diff
@@ -27,11 +27,12 @@
...
 
-import github.GithubObject
+from github.GithubObject import Attribute, NonCompletableGithubObject, NotSet
 
 
-class HookDescription(github.GithubObject.NonCompletableGithubObject):
+class HookDescription(NonCompletableGithubObject):
     """
     This class represents HookDescriptions
     """
```
Comment (EnricoMi) [inline github/HookDescription.py:38]:
```suggestion
class HookDescription(NonCompletableGithubObject):
    """
    This class represents HookDescriptions
    """

    _events: Attribute[list[str]]
    _name: Attribute[str]
    _schema: Attribute[list[list[str]]]
    _supported_events: Attribute[list[str]]
```

File: github/HookDelivery.py [hunk 2]
```diff
@@ -44,86 +44,50 @@ def __repr__(self) -> str:
```
Comment (EnricoMi) [inline github/HookDelivery.py:44]: definition of attributes is missing before `__repr__`:

```python
    _request: Attribute[HookDeliveryRequest]
    _response: Attribute[HookDeliveryResponse]
```

File: github/HookResponse.py [hunk 3]
```diff
@@ -28,10 +28,10 @@
...
 
-import github.GithubObject
+from github.GithubObject import Attribute, NonCompletableGithubObject, NotSet
 
 
-class HookResponse(github.GithubObject.NonCompletableGithubObject):
+class HookResponse(NonCompletableGithubObject):
     """
     This class represents HookResponses
     """
```
Comment (EnricoMi) [inline github/HookResponse.py:37]:
```suggestion
class HookResponse(NonCompletableGithubObject):
    """
    This class represents HookResponses
    """

    _code: Attribute[int]
    _message: Attribute[str]
    _status: Attribute[str]
```

**Review by EnricoMi — CHANGES_REQUESTED** [review]

File: github/BranchProtection.py [hunk 4]
```diff
@@ -35,70 +47,62 @@ class BranchProtection(github.GithubObject.CompletableGithubObject):
...
-        return github.PaginatedList.PaginatedList(
+        return PaginatedList(
             github.NamedUser.NamedUser,
             self._requester,
             self._user_push_restrictions,
             None,
         )
 
-    def get_team_push_restrictions(self):
+    def get_team_push_restrictions(self) -> PaginatedList[Team] | None:
```
[the remark this answers is not in the document]
Comment (trim21) [reply github/BranchProtection.py:87]: yes，it's the new optional.

And also pyupgrade prefer this new union style

File: github/HookDelivery.py (hunk 2 above)
Comment (trim21) [reply github/HookDelivery.py:44]: It's in the method initAttributes

File: github/HookDescription.py (hunk 1 above)
Comment (trim21) [reply github/HookDescription.py:38]: It's in the method initAttributes

File: github/HookResponse.py (hunk 3 above)
Comment (trim21) [reply github/HookResponse.py:37]: It's in the method initAttributes

**Comment by trim21** [comment]
defined instance attribute in `_initAttributes` also works, and don't need to write attributes twice.

File: github/HookDescription.py (hunk 1 above)
Comment (EnricoMi) [reply github/HookDescription.py:38]: It should be in one place for all classes: Either at the beginning of the class, or in the `_initAttributes`. Most classes in this PR do the former, some the latter.

Maybe we move the `_initAttributes` up to be the first method in a class, then we have both: defining types in `_initAttributes` and defining attributes at the beginning (in the first method) of the class.

Comment (trim21) [reply github/HookDescription.py:38]: move _initAttributes to first looks like a good idea

Comment (EnricoMi) [reply github/HookDescription.py:38]: Now we have two ways of typing the attributes, can we stick to one for all classes?

Comment (trim21) [reply github/HookDescription.py:38]: Ok, I'll stick to `_initAttributes` in this PR, and change other existing seprated instance attribute in seprated PR.

File: github/GitRef.py [hunk 5]
```diff
@@ -28,74 +28,71 @@
...
-        :rtype: None
         """
         assert isinstance(sha, str), sha
-        assert force is github.GithubObject.NotSet or isinstance(force, bool), force
-        post_parameters = {
-            "sha": sha,
-        }
-        if force is not github.GithubObject.NotSet:
-            post_parameters["force"] = force
+        assert isinstance(force, (_NotSetType, bool)), force
```
Comment (trim21) [inline github/GitRef.py:88]:
```suggestion
        assert isinstance(force, (_NotSetType, bool)), force
```

File: github/InputFileContent.py [hunk 6]
```diff
@@ -26,32 +26,33 @@
...
 
 
 class InputFileContent:
     """
     This class represents InputFileContents
     """
 
-    def __init__(self, content, new_name=github.GithubObject.NotSet):
+    def __init__(self, content: str, new_name: Opt[str] = NotSet):
         """
```
Comment (trim21) [inline github/InputFileContent.py:41]:
```suggestion
        """
```

File: github/InputFileContent.py [hunk 7]
```diff
@@ -26,32 +26,33 @@
...
         :param content: string
         :param new_name: string
         """
 
         assert isinstance(content, str), content
-        assert new_name is github.GithubObject.NotSet or isinstance(
-            new_name, str
-        ), new_name
+        assert isinstance(new_name, (_NotSetType, str)), new_name
```
Comment (trim21) [inline github/InputFileContent.py:47]:
```suggestion
        assert isinstance(new_name, (_NotSetType, str)), new_name
```

File: github/InputGitAuthor.py [hunk 8]
```diff
@@ -28,15 +28,20 @@
...
 
 class InputGitAuthor:
     """
     This class represents InputGitAuthors
     """
 
-    def __init__(self, name, email, date=github.GithubObject.NotSet):
+    def __init__(
+        self, name: str, email: str, date: Opt[str] = github.GithubObject.NotSet
+    ):
```
Comment (trim21) [inline github/InputGitAuthor.py:44]:
```suggestion
    ):
```

File: github/InputGitTreeElement.py [hunk 9]
```diff
@@ -53,27 +56,23 @@ def __init__(
         assert isinstance(path, str), path
         assert isinstance(mode, str), mode
         assert isinstance(type, str), type
-        assert content is github.GithubObject.NotSet or isinstance(
-            content, str
-        ), content
-        assert (
-            sha is github.GithubObject.NotSet or sha is None or isinstance(sha, str)
-        ), sha
+        assert isinstance(content, (_NotSetType, str)), content
```
Comment (trim21) [inline github/InputGitTreeElement.py:59]:
```suggestion
        assert isinstance(content, (_NotSetType, str)), content
```

**Comment by trim21** [comment]
looks like github pr is broken, some of my commit in this branch don't show up

File: github/InputFileContent.py [hunk 10]
```diff
@@ -26,32 +26,30 @@
...
 
     @property
-    def _identity(self):
-        identity = {
+    def _identity(self) -> dict[str, str]:
+        identity: dict[str, Any] = {
             "content": self.__content,
         }
-        if self.__newName is not github.GithubObject.NotSet:
+        if not is_undefined(self.__newName):
```
Comment (EnricoMi) [inline github/InputFileContent.py:53]:
```suggestion
        if is_defined(self.__newName):
```

File: github/InputGitTreeElement.py [hunk 11]
```diff
@@ -53,27 +56,23 @@ def __init__(
...
-    def _identity(self):
-        identity = {
+    def _identity(self) -> dict[str, Any]:
+        identity: dict[str, Any] = {
             "path": self.__path,
             "mode": self.__mode,
             "type": self.__type,
         }
-        if self.__sha is not github.GithubObject.NotSet:
+        if not is_undefined(self.__sha):
```
Comment (EnricoMi) [inline github/InputGitTreeElement.py:74]:
```suggestion
        if is_defined(self.__sha):
```

File: github/InputGitTreeElement.py [hunk 12]
```diff
@@ -53,27 +56,23 @@ def __init__(
...
             identity["sha"] = self.__sha
-        if self.__content is not github.GithubObject.NotSet:
+        if not is_undefined(self.__content):
```
Comment (EnricoMi) [inline github/InputGitTreeElement.py:76]:
```suggestion
        if is_defined(self.__content):
```

**Review by EnricoMi — APPROVED** [review]
600 fewer lines of code, yay! Minor comment.

**Review by EnricoMi — APPROVED** [review]
LGTM!

**Pull request merged**